### PR TITLE
Support columnar processing for FlatMapCoGroupInPandas[databricks]

### LINKED ISSUE #305: [FEA] Support Effecient data transfers for python UDFs
pyspark supports sending data using arrow in some cases.  Our data is already in arrow format.  We should look at what it would take to support taking our columnar data and sending it across the wire to a python process with as little transformations as possible. In some cases we might even be able to not pull all of the data back from the GPU when we do this.

The class we need to look at is `ArrowEvalPythonExec`.  We should also pay attention to how we do this because we could reuse this for R too if we wanted to.

### PR DESCRIPTION
**Pull request by firestarman**
closes #305

This PR implements the columnar support for the `FlatMapCoGroupInPandasExec`, along with some refactors.

- Moved the GPU Arrow Python runner and its related classes into a separate file.
- Cleaned some boilerplate code. 

***Performance***
- About 6.8 GB Parquet data in local files, containing 100 groups.
- CPU 12 cores, and one GPU (Titan V, with 12GB memory)

| CPU Read + CPU CoGroups  | GPU Read + CPU CoGroups  | GPU Read + GPU CoGroups  |
| ------------ | ------------ | ------------ |
| 103.57  | 89.87  | 81.01  |

Signed-off-by: firestarman <firestarmanllc@gmail.com>

### CODE REVIEW

**Comment by firestarman** [comment]
Draft due to missing the benchmark numbers, working on it now.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
depends on https://github.com/rapidsai/cudf/pull/11883, which is just merged. Waiting for the latest jni jar to run the premerge CI.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Review by jlowe — APPROVED** [review]
Small typo but otherwise lgtm.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuCoGroupedArrowPythonRunner.scala [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import org.apache.spark.api.python.{ChainedPythonFunctions, PythonRDD}
+import org.apache.spark.sql.execution.python.PythonUDFRunner
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.util.Utils
+
+/**
+ * Python UDF Runner for cogrouped UDFs, designed for `GpuFlatMapCoGroupsInPandasExec` only.
+ *
+ * It sends Arrow bathes from two different DataFrames, groups them in Python,
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuCoGroupedArrowPythonRunner.scala:35]: Nice catch, will update this in a following PR for issue #6778.

**Pull request merged**